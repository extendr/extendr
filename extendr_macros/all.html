<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="extendr_macros" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.0-nightly (bf3c6c5be 2024-02-01)" data-channel="nightly" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../extendr_macros/index.html">extendr_macros</a><span class="version">0.6.0</span></h2></div><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#constants">Constants</a></li><li><a href="#functions">Functions</a></li><li><a href="#attributes">Attribute Macros</a></li><li><a href="#derives">Derive Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../extendr_macros/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="call/struct.Call.html">call::Call</a></li><li><a href="extendr_module/struct.Module.html">extendr_module::Module</a></li><li><a href="pairs/struct.Pairs.html">pairs::Pairs</a></li><li><a href="wrappers/struct.ExtendrOptions.html">wrappers::ExtendrOptions</a></li></ul><h3 id="macros">Macros</h3><ul class="all-items"><li><a href="macro.R.html">R</a></li><li><a href="macro.Rraw.html">Rraw</a></li><li><a href="macro.call.html">call</a></li><li><a href="macro.extendr_module.html">extendr_module</a></li><li><a href="macro.list.html">list</a></li><li><a href="macro.pairlist.html">pairlist</a></li></ul><h3 id="attributes">Attribute Macros</h3><ul class="all-items"><li><a href="attr.extendr.html">extendr</a></li></ul><h3 id="derives">Derive Macros</h3><ul class="all-items"><li><a href="derive.IntoDataFrameRow.html">IntoDataFrameRow</a></li><li><a href="derive.IntoRobj.html">IntoRobj</a></li><li><a href="derive.TryFromRobj.html">TryFromRobj</a></li></ul><h3 id="functions">Functions</h3><ul class="all-items"><li><a href="R/fn.R.html">R::R</a></li><li><a href="call/fn.call.html">call::call</a></li><li><a href="dataframe/fn.derive_into_dataframe.html">dataframe::derive_into_dataframe</a></li><li><a href="dataframe/fn.parse_struct.html">dataframe::parse_struct</a></li><li><a href="extendr_function/fn.extendr_function.html">extendr_function::extendr_function</a></li><li><a href="extendr_impl/fn.extendr_impl.html">extendr_impl::extendr_impl</a></li><li><a href="extendr_module/fn.extendr_module.html">extendr_module::extendr_module</a></li><li><a href="list/fn.list.html">list::list</a></li><li><a href="list_struct/fn.derive_into_robj.html">list_struct::derive_into_robj</a></li><li><a href="list_struct/fn.derive_try_from_robj.html">list_struct::derive_try_from_robj</a></li><li><a href="pairlist/fn.pairlist.html">pairlist::pairlist</a></li><li><a href="wrappers/fn.get_doc_string.html">wrappers::get_doc_string</a></li><li><a href="wrappers/fn.get_named_lit.html">wrappers::get_named_lit</a></li><li><a href="wrappers/fn.get_return_type.html">wrappers::get_return_type</a></li><li><a href="wrappers/fn.make_function_wrappers.html">wrappers::make_function_wrappers</a></li><li><a href="wrappers/fn.mangled_type_name.html">wrappers::mangled_type_name</a></li><li><a href="wrappers/fn.sanitize_identifier.html">wrappers::sanitize_identifier</a></li><li><a href="wrappers/fn.translate_actual.html">wrappers::translate_actual</a></li><li><a href="wrappers/fn.translate_formal.html">wrappers::translate_formal</a></li><li><a href="wrappers/fn.translate_meta_arg.html">wrappers::translate_meta_arg</a></li><li><a href="wrappers/fn.translate_to_robj.html">wrappers::translate_to_robj</a></li><li><a href="wrappers/fn.type_name.html">wrappers::type_name</a></li></ul><h3 id="constants">Constants</h3><ul class="all-items"><li><a href="wrappers/constant.META_PREFIX.html">wrappers::META_PREFIX</a></li><li><a href="wrappers/constant.WRAP_PREFIX.html">wrappers::WRAP_PREFIX</a></li></ul></section></div></main></body></html>