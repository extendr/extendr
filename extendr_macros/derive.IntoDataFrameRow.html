<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Enable the construction of dataframes from arrays of structures."><title>IntoDataFrameRow in extendr_macros - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-dd39b87e5fcfba68.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="extendr_macros" data-themes="" data-resource-suffix="" data-rustdoc-version="1.80.0-nightly (6e1d94708 2024-05-10)" data-channel="nightly" data-search-js="search-d52510db62a78183.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc derive"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../extendr_macros/index.html">extendr_macros</a><span class="version">0.6.0</span></h2></div><div class="sidebar-elems"></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Derive Macro <a href="index.html">extendr_macros</a>::<wbr><a class="derive" href="#">IntoDataFrameRow</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/extendr_macros/lib.rs.html#300-302">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>#[derive(IntoDataFrameRow)]</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Enable the construction of dataframes from arrays of structures.</p>
<h2 id="example"><a class="doc-anchor" href="#example">§</a>Example</h2>
<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="kw">use </span>extendr_api::prelude::<span class="kw-2">*</span>;

<span class="attr">#[derive(Debug, IntoDataFrameRow)]
</span><span class="kw">struct </span>MyStruct {
    x: i32,
    y: String,
}

<span class="kw">let </span>v = <span class="macro">vec!</span>[MyStruct { x: <span class="number">0</span>, y: <span class="string">"abc"</span>.into() }, MyStruct { x: <span class="number">1</span>, y: <span class="string">"xyz"</span>.into() }];
<span class="kw">let </span>df = v.into_dataframe()<span class="question-mark">?</span>;

<span class="macro">assert!</span>(df.inherits(<span class="string">"data.frame"</span>));
<span class="macro">assert_eq!</span>(df[<span class="number">0</span>], <span class="macro">r!</span>([<span class="number">0</span>, <span class="number">1</span>]));
<span class="macro">assert_eq!</span>(df[<span class="number">1</span>], <span class="macro">r!</span>([<span class="string">"abc"</span>, <span class="string">"xyz"</span>]));</code></pre></div>
</div></details></section></div></main></body></html>