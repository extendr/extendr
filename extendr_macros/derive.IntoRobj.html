<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Derives an implementation of `From&lt;Struct&gt; for Robj` and `From&lt;&amp;Struct&gt; for Robj` on this struct."><title>IntoRobj in extendr_macros - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-f40c346f39d9abc1.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="extendr_macros" data-themes="" data-resource-suffix="" data-rustdoc-version="1.72.0-nightly (065a1f5df 2023-06-21)" data-search-js="search-95c92dd01058facf.js" data-settings-js="settings-de11bff964e9d4e5.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../static.files/storage-62ce34ea385b278a.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-190c35055d2a8300.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc derive"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../extendr_macros/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../extendr_macros/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><div class="sidebar-elems"><h2><a href="index.html">In extendr_macros</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Derive Macro <a href="index.html">extendr_macros</a>::<wbr><a class="derive" href="#">IntoRobj</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/extendr_macros/lib.rs.html#239-241">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>#[derive(IntoRobj)]</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Derives an implementation of <code>From&lt;Struct&gt; for Robj</code> and <code>From&lt;&amp;Struct&gt; for Robj</code> on this struct.</p>
<p>This allows the struct to be converted to a named list in R,
where the list names correspond to the field names of the Rust struct.</p>
<h2 id="examples"><a href="#examples">Examples</a></h2>
<p>In the below example, <code>converted</code> contains an R list object with the same fields as the
<code>Foo</code> struct.</p>

<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="kw">use </span>extendr_api::prelude::<span class="kw-2">*</span>;
<span class="kw">use </span>extendr_macros::IntoRobj;

<span class="attr">#[derive(IntoRobj)]
</span><span class="kw">struct </span>Foo {
    a: u32,
    b: String
}
<span class="kw">let </span>converted: Robj = Foo {
    a: <span class="number">5</span>,
    b: String::from(<span class="string">&quot;bar&quot;</span>)
}.into();
<span class="macro">assert_eq!</span>(converted, <span class="macro">R!</span>(<span class="string">r&quot;list(a=5, b=&#39;bar&#39;)&quot;</span>)<span class="question-mark">?</span>);</code></pre></div>
<h2 id="details"><a href="#details">Details</a></h2>
<p>Note, the <code>From&lt;Struct&gt; for Robj</code> behaviour is different from what is obtained by applying the standard <code>#[extendr]</code> macro
to an <code>impl</code> block. The <code>#[extendr]</code> behaviour returns to R a <strong>pointer</strong> to Rust memory, and generates wrapper functions for calling
Rust functions on that pointer. The implementation from <code>#[derive(IntoRobj)]</code> actually converts the Rust structure
into a native R list, which allows manipulation and access to internal fields, but it’s a one-way conversion,
and converting it back to Rust will produce a copy of the original struct.</p>
</div></details></section></div></main></body></html>