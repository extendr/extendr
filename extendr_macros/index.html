<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Macros for generating wrappers for rust functions."><title>extendr_macros - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-f40c346f39d9abc1.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="extendr_macros" data-themes="" data-resource-suffix="" data-rustdoc-version="1.72.0-nightly (065a1f5df 2023-06-21)" data-search-js="search-95c92dd01058facf.js" data-settings-js="settings-de11bff964e9d4e5.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../static.files/storage-62ce34ea385b278a.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-190c35055d2a8300.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../extendr_macros/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../extendr_macros/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate extendr_macros</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.4.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#attributes">Attribute Macros</a></li><li><a href="#derives">Derive Macros</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press â€˜Sâ€™ to search, â€˜?â€™ for more optionsâ€¦" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">extendr_macros</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/extendr_macros/lib.rs.html#1-267">source</a> Â· <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Macros for generating wrappers for rust functions.</p>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="R/index.html" title="mod extendr_macros::R">R</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="call/index.html" title="mod extendr_macros::call">call</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="dataframe/index.html" title="mod extendr_macros::dataframe">dataframe</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="extendr_function/index.html" title="mod extendr_macros::extendr_function">extendr_function</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="extendr_impl/index.html" title="mod extendr_macros::extendr_impl">extendr_impl</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="extendr_module/index.html" title="mod extendr_macros::extendr_module">extendr_module</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="list/index.html" title="mod extendr_macros::list">list</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="list_struct/index.html" title="mod extendr_macros::list_struct">list_struct</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="pairlist/index.html" title="mod extendr_macros::pairlist">pairlist</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li><li><div class="item-name"><a class="mod" href="pairs/index.html" title="mod extendr_macros::pairs">pairs</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div><div class="desc docblock-short">Internal module for parsing R-like variadic arguments.</div></li><li><div class="item-name"><a class="mod" href="wrappers/index.html" title="mod extendr_macros::wrappers">wrappers</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </div></li></ul><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.R.html" title="macro extendr_macros::R">R</a></div><div class="desc docblock-short">Execute R code by parsing and evaluating tokens.</div></li><li><div class="item-name"><a class="macro" href="macro.Rraw.html" title="macro extendr_macros::Rraw">Rraw</a></div><div class="desc docblock-short">Execute R code by parsing and evaluating tokens
but without expanding parameters.</div></li><li><div class="item-name"><a class="macro" href="macro.call.html" title="macro extendr_macros::call">call</a></div><div class="desc docblock-short">Call a function or primitive defined by a text expression with arbitrary parameters.
This currently works by parsing and evaluating the string in R, but will probably acquire
some shortcuts for simple expessions, for example by caching symbols and constant values.</div></li><li><div class="item-name"><a class="macro" href="macro.extendr_module.html" title="macro extendr_macros::extendr_module">extendr_module</a></div><div class="desc docblock-short">Define a module and export symbols to R
Example:</div></li><li><div class="item-name"><a class="macro" href="macro.list.html" title="macro extendr_macros::list">list</a></div><div class="desc docblock-short">Create a List R object from a list of name-value pairs.</div></li><li><div class="item-name"><a class="macro" href="macro.pairlist.html" title="macro extendr_macros::pairlist">pairlist</a></div><div class="desc docblock-short">Create a Pairlist R object from a list of name-value pairs.</div></li></ul><h2 id="attributes" class="small-section-header"><a href="#attributes">Attribute Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.extendr.html" title="attr extendr_macros::extendr">extendr</a></div></li></ul><h2 id="derives" class="small-section-header"><a href="#derives">Derive Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="derive" href="derive.IntoDataFrameRow.html" title="derive extendr_macros::IntoDataFrameRow">IntoDataFrameRow</a></div><div class="desc docblock-short">Enable the construction of dataframes from arrays of structures.</div></li><li><div class="item-name"><a class="derive" href="derive.IntoRobj.html" title="derive extendr_macros::IntoRobj">IntoRobj</a></div><div class="desc docblock-short">Derives an implementation of <code>From&lt;Struct&gt; for Robj</code> and <code>From&lt;&amp;Struct&gt; for Robj</code> on this struct.</div></li><li><div class="item-name"><a class="derive" href="derive.TryFromRobj.html" title="derive extendr_macros::TryFromRobj">TryFromRobj</a></div><div class="desc docblock-short">Derives an implementation of <code>TryFrom&lt;Robj&gt; for Struct</code> and <code>TryFrom&lt;&amp;Robj&gt; for Struct</code> on this struct.</div></li></ul></section></div></main></body></html>