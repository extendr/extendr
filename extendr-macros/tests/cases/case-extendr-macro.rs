use extendr_macros::extendr;
use extendr_macros::{IntoRobj, TryFromRobj};

#[extendr(foo = true)]
fn foo() {}

#[extendr(use_try_from = 1)]
fn foo() {}

#[extendr(r_name = 1)]
fn foo() {}

#[extendr(mod_name = 1)]
fn foo() {}

#[extendr(use_rng = 1)]
fn foo() {}

// impl -----------------------------------------

struct FooStruct {}

#[extendr]
impl FooStruct {
    fn nonref_self(self) {}
}

impl FooStruct {
    #[extendr]
    fn misplaced_macro(&self) {}
}

#[extendr]
default impl FooStruct {}

#[extendr]
unsafe impl FooStruct {}

#[extendr]
impl<const N: usize> FooStruct {}

struct FooStructWithParam<A> {
    a: A,
}

#[extendr]
impl<A> FooStructWithParam<A> {}

#[extendr]
impl FooStructWithParam<A> where A: usize {}

// derive ---------------------------------------

#[derive(TryFromRobj)]
enum FooEnum1 {
    A,
    B,
    C,
}

#[derive(IntoRobj)]
enum FooEnum2 {
    A,
    B,
    C,
}

fn main() {}
