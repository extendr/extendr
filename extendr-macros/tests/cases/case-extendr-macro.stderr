error: Unexpected key
 --> tests/cases/case-extendr-macro.rs:4:11
  |
4 | #[extendr(foo = true)]
  |           ^^^

error: unexpected end of input, `r_name` must be a string literal
 --> tests/cases/case-extendr-macro.rs:7:1
  |
7 | #[extendr(r_name = 1)]
  | ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, `mod_name` must be a string literal
  --> tests/cases/case-extendr-macro.rs:10:1
   |
10 | #[extendr(mod_name = 1)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of input, `use_rng` must be `true` or `false`
  --> tests/cases/case-extendr-macro.rs:13:1
   |
13 | #[extendr(use_rng = 1)]
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected &self or &mut self
  --> tests/cases/case-extendr-macro.rs:22:20
   |
22 |     fn nonref_self(self) {}
   |                    ^^^^

error: found &self in non-impl function - have you missed the #[extendr] before the impl?
  --> tests/cases/case-extendr-macro.rs:27:24
   |
27 |     fn misplaced_macro(&self) {}
   |                        ^^^^^

error: default not allowed in #[extendr] impl
  --> tests/cases/case-extendr-macro.rs:31:1
   |
31 | default impl FooStruct {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

error: unsafe not allowed in #[extendr] impl
  --> tests/cases/case-extendr-macro.rs:34:1
   |
34 | unsafe impl FooStruct {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^

error: const params not allowed in #[extendr] impl
  --> tests/cases/case-extendr-macro.rs:37:1
   |
37 | impl<const N: usize> FooStruct {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: type params not allowed in #[extendr] impl
  --> tests/cases/case-extendr-macro.rs:44:1
   |
44 | impl<A> FooStructWithParam<A> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: where clause not allowed in #[extendr] impl
  --> tests/cases/case-extendr-macro.rs:47:1
   |
47 | impl FooStructWithParam<A> where A: usize {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Only struct is supported
  --> tests/cases/case-extendr-macro.rs:52:1
   |
52 | / enum FooEnum1 {
53 | |     A,
54 | |     B,
55 | |     C,
56 | | }
   | |_^

error: Only struct is supported
  --> tests/cases/case-extendr-macro.rs:59:1
   |
59 | / enum FooEnum2 {
60 | |     A,
61 | |     B,
62 | |     C,
63 | | }
   | |_^
